decryptor/postgresql/data_encoder.go

```diff
@@ -0,0 +1,255 @@
+			if setting.GetDefaultDataValue() != nil {
+				newVal := setting.GetDefaultDataValue()
+				logger.WithField("data", string(data)).WithField("default", *newVal).Debugln("Change with default")
+				return ctx, []byte(*newVal), nil
+			}
+		}
+	case common2.EncryptedType_Bytes:
+		if !base.IsDecryptedFromContext(ctx) {
+			if setting.GetDefaultDataValue() != nil {
+				newVal := setting.GetDefaultDataValue()
+				binValue, err := base64.StdEncoding.DecodeString(*newVal)
+				if err != nil {
+					return ctx, data, err
+				}
+				// override and encode at end of function
+				data = binValue
+			}
+		}
+	case common2.EncryptedType_Int32, common2.EncryptedType_Int64:
+		_, err := strconv.ParseInt(string(data), 10, 64)
+		// if it's valid string literal and decrypted, return as is
+		if err == nil {
+			return ctx, data, nil
+		}
+		// if it's encrypted binary, then it is binary array that is invalid int literal
+		if !base.IsDecryptedFromContext(ctx) {
+			if setting.GetDefaultDataValue() != nil {
+				newVal := setting.GetDefaultDataValue()
+				logger.Debugln("Return default value")
+				return ctx, []byte(*newVal), nil
+			}
+		}
+		return ctx, data, nil
+	}
+	if utils.IsPrintablePostgresqlString(data) {
+		return ctx, data, nil
+	}
+	return ctx, utils.PgEncodeToHex(data), nil
+}
+
+// decodeText converts data from text format for decryptors/de-tokenizers according to ColumnEncryptionSetting
+// hex/octal binary -> raw binary data
+func (p *PgSQLDataEncoderProcessor) decodeText(ctx context.Context, data []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	if config.IsBinaryDataOperation(setting) {
+		// decryptor operates over blobs so all data types will be encrypted as hex/octal string values that we should
+		// decode before decryption
+		decodedData, err := utils.DecodeEscaped(data)
+		if err != nil {
+			logger.WithError(err).Errorln("Can't decode binary data for decryption")
+			return ctx, data, nil
+		}
+		return ctx, decodedData, nil
+	}
+	// all other non-binary data should be valid SQL literals like integers or strings and Acra works with them as is
+	return ctx, data, nil
+}
+
+// OnColumn encode binary value to text and back. Should be before and after tokenizer processor
+func (p *PgSQLDataEncoderProcessor) OnColumn(ctx context.Context, data []byte) (context.Context, []byte, error) {
+	columnSetting, ok := encryptor.EncryptionSettingFromContext(ctx)
+	if !ok {
+		// for case when data encrypted with acrastructs on app's side and used without any encryption setting
+		columnSetting = &config.BasicColumnEncryptionSetting{}
+	}
+	logger := logging.GetLoggerFromContext(ctx)
+	columnInfo, ok := base.ColumnInfoFromContext(ctx)
+	if !ok {
+		logger.WithField("processor", "PgSQLDataEncoderProcessor").Warningln("No column info in ctx")
+		// we can't do anything
+		return ctx, data, nil
+	}
+
+	if p.mode == DataEncoderModeEncode {
```

Comment:
Do you think to split this logic into two different parts?

I mean something like this.

```

type BasePgSQLDataEncoderProcessor struct {}

func (p *BasePgSQLDataEncoderProcessor) encodeBinary()

func (p *BasePgSQLDataEncoderProcessor) decodeBinary()

func (p *BasePgSQLDataEncoderProcessor) encodeText()

func (p *BasePgSQLDataEncoderProcessor) decodeText()





type DecodePgSQLDataEncoderProcessor struct {
     BasePgSQLDataEncoderProcessor
}
// with it own decode logic
func (p *DecodePgSQLDataEncoderProcessor) OnColumn()



type EncodePgSQLDataEncoderProcessor struct {
     BasePgSQLDataEncoderProcessor
}

//with it own encode logic
func (p *EncodePgSQLDataEncoderProcessor) OnColumn()
```

We can decouple this logic from one place which can be easier to support in future. What do you think?